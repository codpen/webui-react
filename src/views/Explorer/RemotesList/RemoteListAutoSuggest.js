import Autosuggest from 'react-autosuggest';
import React from "react";
import PropTypes from "prop-types";

// Teach Autosuggest how to calculate suggestions for any given input value.
const getSuggestions = (config, value) => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;

    if (inputLength === 0) {
        return config;

    }

    return inputLength === 0 ? [] : config.filter(lang =>
        lang.toLowerCase().slice(0, inputLength) === inputValue
    );
};

// When suggestion is clicked, Autosuggest needs to populate the input
// based on the clicked suggestion. Teach Autosuggest how to calculate the
// input value for every given suggestion.
const getSuggestionValue = suggestion => suggestion;

// Use your imagination to render suggestions.
const renderSuggestion = suggestion => (
    <div>
        <i className="fa fa-lg fa-hdd-o"/> {suggestion}
    </div>
);

class RemoteListAutoSuggest extends React.Component {
    constructor(props) {
        super(props);

        // Autosuggest is a controlled component.
        // This means that you need to provide an input value
        // and an onChange handler that updates this value (see below).
        // Suggestions also need to be provided to the Autosuggest,
        // and they are initially empty because the Autosuggest is closed.
        this.state = {
            suggestions: []
        };
    }


    // Autosuggest will call this function every time you need to update suggestions.
    // You already implemented this logic above, so just use it.
    onSuggestionsFetchRequested = ({value}) => {
        this.setState({
            suggestions: getSuggestions(this.props.suggestions, value)
        });
    };

    // Autosuggest will call this function every time you need to clear suggestions.
    onSuggestionsClearRequested = () => {
        this.setState({
            suggestions: []
        });
    };

    render() {
        const {value, onChange} = this.props;


        // Autosuggest will pass through all these props to the input.
        const inputProps = {
            placeholder: 'Type the name of remote you want to open',
            value: value,
            onChange: onChange
        };

        // Finally, render it!
        return (
            <Autosuggest
                suggestions={this.state.suggestions}
                onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
                onSuggestionsClearRequested={this.onSuggestionsClearRequested}
                getSuggestionValue={getSuggestionValue}
                renderSuggestion={renderSuggestion}
                alwaysRenderSuggestions={true}
                highlightFirstSuggestion={true}
                inputProps={inputProps}
                style={{width:"100%"}}
            />
        );
    }
}

RemoteListAutoSuggest.propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    suggestions: PropTypes.array.isRequired
};

export default RemoteListAutoSuggest;