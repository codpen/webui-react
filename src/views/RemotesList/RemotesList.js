import React from 'react';
import "../../utils/Global";
import axiosInstance from "../../utils/API";
import {Button, Card, CardBody, Col, Row} from "reactstrap";

class RemotesList extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            remotes: [],
            isEmpty: true
        };
    }

    async updateRemoteList() {
        /*
        {
            "remotes": [
                "eventsfunk",
                "mydrive"
            ]
        }
        */
        try {
            let res = await axiosInstance.post("/config/listremotes");
            // console.log("Remote list response:");
            // console.log(res, res.data.remotes.length);
            if (res.data.remotes.length === 0) {
                this.setState({isEmpty: true});
            } else {
                this.setState({remotes: res.data.remotes, isEmpty: false});
            }

        } catch (e) {
            console.log(`Error loading remotes: ${e}`)
        }
    }

    componentDidMount() {
        // TODO: Get remote List via config/listremotes
        this.updateRemoteList();
    }


    render() {
        const {isEmpty, remotes} = this.state;
        const {updateRemoteNameHandle} = this.props;
        if (isEmpty) {
            return (
                <div>
                    Add some remotes to see them here <span role="img" aria-label="sheep">üêë</span>.
                </div>);
        } else {
            let remotesMap = remotes.map((item, idx) => {
                return (
                    <Col key={item} xs={12} sm={6} lg={3}>
                        <Card>
                            <CardBody>
                                <p><strong>Name:</strong> {item}
                                    <Button color={"info"} onClick={() => updateRemoteNameHandle(item)}>Open</Button>
                                </p>
                            </CardBody>
                        </Card>
                    </Col>
                )
            });
            return (
                <Row>
                    {remotesMap}
                </Row>
            );
        }
    }
}

export default RemotesList;
