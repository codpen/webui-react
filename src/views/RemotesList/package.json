{
  "name": "RemotesList",
  "version": "0.0.0",
  "private": true,
  "main": "RemotesList.js"
}
