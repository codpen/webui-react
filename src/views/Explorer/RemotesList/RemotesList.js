import React from 'react';
import RemoteListAutoSuggest from "./RemoteListAutoSuggest";
import {connect} from "react-redux";
import {getFsInfo, getRemoteNames} from "../../../actions/explorerActions";
import PropTypes from 'prop-types'
import {changeRemoteName} from "../../../actions/explorerStateActions";
import {Button, Col, Form, Row} from "reactstrap";

class RemotesList extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            isEmpty: false,
            remoteName: props.remoteName,
            openEnabled: false,
            openButtonText: "Open"
        };
    }

    componentDidMount() {

        this.props.getRemoteNames();
    }

    shouldUpdateRemoteName = (event, {newValue}) => {
        if (newValue.indexOf('/') === 0) {
            this.setState({
                remoteName: newValue,
                openButtonText: "Open local path",

            });
        } else {
            this.setState({
                remoteName: newValue,
                openButtonText: "Open"
            });
        }


    };

    openRemote = () => {
        const {changeRemoteName, containerID} = this.props;
        const {remoteName} = this.state;
        changeRemoteName(containerID, remoteName);

        this.props.getFsInfo(remoteName);

    };


    render() {
        const {isEmpty, remoteName} = this.state;
        const {remotes} = this.props;
        const {hasError} = this.props;
        // const {updateRemoteNameHandle} = this.props;

        if (hasError) {
            return (
                <div>
                    Error loading remotes. Please try again.
                </div>
            )
        } else if (isEmpty) {
            return (
                <div>
                    Add some remotes to see them here <span role="img" aria-label="sheep">üêë</span>.
                </div>);
        } else {

            return (
                <Form onSubmit={() => this.openRemote()}>
                    <Row>

                        <Col xs={12} sm={10} lg={8}>
                            <RemoteListAutoSuggest value={remoteName} onChange={this.shouldUpdateRemoteName}
                                                   suggestions={remotes}/>
                        </Col>
                        <Col xs={12} sm={2} lg={2}>

                            <Button className={"btn-lg"} color="success"
                                    type="submit">{this.state.openButtonText}</Button>
                        </Col>

                    </Row>
                </Form>

            );
        }
    }
}

const mapStateToProps = (state, ownProps) => ({
    remotes: state.remote.remotes,
    hasError: false,
    error: state.remote.error,
    currentPath: state.explorer.currentPaths[ownProps.containerID],
});

const propTypes = {
    remotes: PropTypes.array.isRequired,
    error: PropTypes.object,
    hasError: PropTypes.bool,
    containerID: PropTypes.string.isRequired,

    currentPath: PropTypes.shape({
        remoteName: PropTypes.string.isRequired,
        remotePath: PropTypes.string.isRequired
    })

};


const defaultProps = {};

RemotesList.propTypes = propTypes;
RemotesList.defaultProps = defaultProps;


export default connect(mapStateToProps, {
    getRemoteNames,
    getFsInfo,
    changeRemoteName,

})(RemotesList);
