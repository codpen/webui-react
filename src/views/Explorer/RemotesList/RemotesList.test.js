import React from 'react';
import {testStore} from "../../../../Utils";
import {shallow} from "enzyme";
import {Provider} from "react-redux";
import RemotesList from "./RemotesList";


const setUp = (props = {}) => {
    const component = shallow(
        <Provider store={testStore()}>
            <RemotesList/>
        </Provider>
    );
    return component;
}

describe('Bandwidth Status Card', function () {

    let wrapper;
    beforeEach(() => {
        const props = {};
        wrapper = setUp(props)
    });

    it('should render without crashing', function () {
        expect(wrapper).toHaveLength(1)
    });
});