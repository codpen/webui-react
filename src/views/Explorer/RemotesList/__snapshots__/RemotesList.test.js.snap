// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bandwidth Status Card should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <Connect(RemotesList) />
</ContextProvider>
`;
